File: Data Structures & Algorithms/plus-one/submission-12.cs
class Solution {
    /**
     * @param {number[]} digits
     * @return {number[]}
     */
    plusOne(digits) {
        let one = 1;
        let i = 0;
        digits.reverse();

        while (one) {
            if (i < digits.length) {
                if (digits[i] === 9) {
                    digits[i] = 0;
                } else {
                    digits[i] += 1;
                    one = 0;
                }
            } else {
                digits.push(one);
                one = 0;
            }
            i++;
        }
        digits.reverse();
        return digits;
    }
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-12.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @return {number}
     */
    longestIncreasingPath(matrix) {
        const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
        const ROWS = matrix.length, COLS = matrix[0].length;

        const dfs = (r, c, prevVal) => {
            if (r < 0 || r >= ROWS || c < 0 || 
                c >= COLS || matrix[r][c] <= prevVal) {
                return 0;
            }

            let res = 1;
            for (let d of directions) {
                res = Math.max(res, 1 + dfs(r + d[0], 
                                        c + d[1], matrix[r][c]));
            }
            return res;
        };

        let LIP = 0;
        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                LIP = Math.max(LIP, dfs(r, c, -Infinity));
            }
        }
        return LIP;
    }
}